File: ios-app/Packages/BayitVoice/Sources/BayitVoice/VoiceWebSocketClient.swift
import BayitCore
import Foundation

/// WebSocket client for the streaming voice pipeline.
///
/// Ported from shared/services/voiceWebSocketHandler.ts.
/// Uses URLSessionWebSocketTask (iOS 13+) for native WebSocket support.
/// Handles message routing, keep-alive pings, and reconnection.
public actor VoiceWebSocketClient {

    private var webSocketTask: URLSessionWebSocketTask?
    private var pingTask: Task<Void, Never>?
    private var receiveTask: Task<Void, Never>?
    private let session: URLSession
    private let logger = BayitLogger(category: "VoiceWebSocket")

    private let baseURL: URL
    private let authTokenProvider: @Sendable () async -> String?

    /// Callback for incoming messages, set by the orchestrator.
    private var onMessage: (@Sendable (VoiceWSIncoming) -> Void)?

    /// Ping interval matching voiceWebSocketHandler.ts (30 seconds)
    private let pingInterval: TimeInterval = 30

    public init(
        baseURL: URL,
        authTokenProvider: @escaping @Sendable () async -> String?
    ) {
        self.baseURL = baseURL
        self.authTokenProvider = authTokenProvider
        self.session = URLSession(configuration: .default)
    }

    // MARK: - Connection

    /// Connect to the voice WebSocket endpoint.
    ///
    /// - Parameters:
    ///   - language: Language code for the session (e.g. "en", "he")
    ///   - onMessage: Callback for incoming messages
    public func connect(
        language: String,
        onMessage: @escaping @Sendable (VoiceWSIncoming) -> Void
    ) async {
        // Close any existing connection first
        await disconnect()

        self.onMessage = onMessage

        let wsURL = baseURL.appendingPathComponent("/ws/voice")
        var request = URLRequest(url: wsURL)

        if let token = await authTokenProvider() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.setValue(language, forHTTPHeaderField: "Accept-Language")

        let task = session.webSocketTask(with: request)
        webSocketTask = task
        task.resume()

        logger.info("WebSocket connecting", context: ["url": wsURL.absoluteString])

        startReceiving()
        startPingLoop()
    }

    /// Disconnect and clean up.
    public func disconnect() {
        pingTask?.cancel()
        pingTask = nil
        receiveTask?.cancel()
        receiveTask = nil

        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        onMessage = nil

        logger.info("WebSocket disconnected")
    }

    // MARK: - Send Messages

    /// Send an audio chunk (base64-encoded).
    public func sendAudio(_ audioData: Data) async {
        let base64 = audioData.base64EncodedString()
        let message: [String: Any] = ["type": "audio", "data": base64]
        await sendJSON(message)
    }

    /// Commit the current transcript (finalize).
    public func sendCommit() async {
        await sendJSON(["type": "commit"])
    }

    /// Cancel the current processing.
    public func sendCancel() async {
        await sendJSON(["type": "cancel"])
    }

    // MARK: - Receiving

    private func startReceiving() {
        receiveTask = Task { [weak self] in
            while !Task.isCancelled {
                guard let self, let ws = await self.webSocketTask else { break }
                do {
                    let message = try await ws.receive()
                    await self.handleMessage(message)
                } catch {
                    if !Task.isCancelled {
                        await self.logger.error("WebSocket receive error", error: error)
                        await self.onMessage?(.error(error.localizedDescription))
                    }
                    break
                }
            }
        }
    }

    private func handleMessage(_ message: URLSessionWebSocketTask.Message) {
        switch message {
        case .string(let text):
            parseTextMessage(text)
        case .data(let data):
            if let text = String(data: data, encoding: .utf8) {
                parseTextMessage(text)
            }
        @unknown default:
            break
        }
    }

    /// Parse and route incoming JSON messages.
    /// Validates message types matching voiceWebSocketHandler.ts whitelist.
    private func parseTextMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else {
            logger.warning("Invalid WebSocket message format")
            return
        }

        switch type {
        case "transcript_partial":
            if let transcript = json["text"] as? String {
                onMessage?(.transcriptPartial(transcript))
            }

        case "transcript_final":
            if let transcript = json["text"] as? String {
                onMessage?(.transcriptFinal(transcript))
            }

        case "llm_chunk":
            if let chunk = json["text"] as? String {
                onMessage?(.llmChunk(chunk))
            }

        case "tts_audio":
            if let base64 = json["data"] as? String,
               let audioData = Data(base64Encoded: base64) {
                onMessage?(.ttsAudio(audioData))
            }

        case "intent_action":
            if let intentStr = json["intent"] as? String,
               let actionText = json["text"] as? String,
               let actionDict = json["action"] as? [String: Any],
               let actionType = actionDict["type"] as? String {
                let intent = VoiceIntentType(rawValue: intentStr) ?? .unknown
                let payload = (actionDict["payload"] as? [String: Any])?.compactMapValues {
                    AnyCodable($0)
                }
                let action = VoiceAction(type: actionType, payload: payload)
                onMessage?(.intentAction(intent, actionText, action))
            }

        case "complete":
            onMessage?(.complete)

        case "cancelled":
            onMessage?(.cancelled)

        case "error":
            let errorMsg = json["message"] as? String ?? "Unknown server error"
            onMessage?(.error(errorMsg))

        case "pong":
            onMessage?(.pong)

        default:
            logger.warning("Unknown WebSocket message type", context: ["type": type])
        }
    }

    // MARK: - Ping Keep-Alive

    private func startPingLoop() {
        pingTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(self?.pingInterval ?? 30))
                guard !Task.isCancelled else { break }
                await self?.sendJSON(["type": "ping"])
            }
        }
    }

    // MARK: - Helpers

    private func sendJSON(_ dict: [String: Any]) async {
        guard let data = try? JSONSerialization.data(withJSONObject: dict),
              let text = String(data: data, encoding: .utf8) else {
            return
        }
        do {
            try await webSocketTask?.send(.string(text))
        } catch {
            logger.error("WebSocket send failed", error: error)
        }
    }
}

File: android-app/core/core-voice/src/main/java/tv/bayit/plus/core/voice/VoiceWebSocketClient.kt
package tv.bayit.plus.core.voice

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import tv.bayit.plus.core.common.logging.BayitLogger
import tv.bayit.plus.core.network.NetworkConfiguration
import javax.inject.Inject
import javax.inject.Singleton

/** OkHttp WebSocket client for the streaming voice pipeline. Port of iOS VoiceWebSocketClient. */
@Singleton
class VoiceWebSocketClient @Inject constructor(
    private val okHttpClient: OkHttpClient,
    private val json: Json,
    private val networkConfig: NetworkConfiguration,
    private val logger: BayitLogger,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val mutex = Mutex()
    private var webSocket: WebSocket? = null
    private var pingJob: Job? = null
    private val _messages = MutableSharedFlow<VoiceWSIncoming>(extraBufferCapacity = 64)
    val messages: SharedFlow<VoiceWSIncoming> = _messages
    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected

    private val listener = object : WebSocketListener() {
        override fun onOpen(ws: WebSocket, response: Response) {
            _isConnected.value = true
            logger.info("Voice WebSocket connected", mapOf("code" to "${response.code}"))
            startPingLoop()
        }
        override fun onMessage(ws: WebSocket, text: String) { parseAndEmit(text) }
        override fun onClosing(ws: WebSocket, code: Int, reason: String) { ws.close(code, reason) }
        override fun onClosed(ws: WebSocket, code: Int, reason: String) {
            _isConnected.value = false
            logger.info("Voice WebSocket closed", mapOf("code" to "$code"))
        }
        override fun onFailure(ws: WebSocket, t: Throwable, response: Response?) {
            _isConnected.value = false
            logger.error("Voice WebSocket failure", t, mapOf("response" to "${response?.code}"))
            _messages.tryEmit(VoiceWSIncoming.Error(t.message ?: "Connection failure"))
        }
    }

    suspend fun connect(url: String, token: String, language: String) {
        mutex.withLock {
            closeInternal()
            val request = Request.Builder()
                .url(url)
                .header("Authorization", "Bearer $token")
                .header("Accept-Language", language)
                .build()
            webSocket = okHttpClient.newWebSocket(request, listener)
            logger.info("Voice WebSocket connecting", mapOf("url" to url, "lang" to language))
        }
    }

    suspend fun sendAudio(base64Data: String) = sendJson("""{"type":"audio","data":"$base64Data"}""")
    suspend fun sendCommit() = sendJson("""{"type":"commit"}""")
    suspend fun sendCancel() = sendJson("""{"type":"cancel"}""")

    suspend fun close() { mutex.withLock { closeInternal() } }

    private fun closeInternal() {
        pingJob?.cancel()
        pingJob = null
        webSocket?.close(NORMAL_CLOSURE_CODE, NORMAL_CLOSURE_REASON)
        webSocket = null
        _isConnected.value = false
    }

    private suspend fun sendJson(text: String) {
        mutex.withLock {
            if (webSocket?.send(text) != true) {
                logger.warning("Voice WebSocket send failed: not connected")
            }
        }
    }

    private fun startPingLoop() {
        pingJob?.cancel()
        pingJob = scope.launch {
            val intervalMs = networkConfig.webSocketPingIntervalDuration.inWholeMilliseconds
            while (_isConnected.value) {
                delay(intervalMs)
                if (_isConnected.value) mutex.withLock { webSocket?.send("""{"type":"ping"}""") }
            }
        }
    }

    private fun parseAndEmit(text: String) {
        val obj = runCatching { json.parseToJsonElement(text).jsonObject }.getOrNull()
        if (obj == null) { logger.warning("Invalid WebSocket message format"); return }
        val type = obj["type"]?.jsonPrimitive?.contentOrNull
        if (type == null) { logger.warning("WebSocket message missing type field"); return }
        routeMessage(type, obj)?.let { _messages.tryEmit(it) }
    }

    private fun routeMessage(type: String, obj: JsonObject): VoiceWSIncoming? = when (type) {
        "transcript_partial" -> strVal(obj, "text")?.let { VoiceWSIncoming.TranscriptPartial(it) }
        "transcript_final" -> strVal(obj, "text")?.let { VoiceWSIncoming.TranscriptFinal(it) }
        "llm_chunk" -> strVal(obj, "text")?.let { VoiceWSIncoming.LlmChunk(it) }
        "tts_audio" -> decodeBase64Val(obj, "data")?.let { VoiceWSIncoming.TtsAudio(it) }
        "intent_action" -> parseIntentAction(obj)
        "complete" -> VoiceWSIncoming.Complete(strVal(obj, "conversation_id"))
        "cancelled" -> VoiceWSIncoming.Cancelled
        "error" -> VoiceWSIncoming.Error(strVal(obj, "message") ?: "Unknown server error")
        "pong" -> VoiceWSIncoming.Pong
        else -> { logger.warning("Unknown WebSocket message type", mapOf("type" to type)); null }
    }

    private fun parseIntentAction(obj: JsonObject): VoiceWSIncoming.IntentAction? {
        val intentStr = strVal(obj, "intent") ?: return null
        val text = strVal(obj, "text") ?: return null
        val actionObj = obj["action"]?.jsonObject ?: return null
        val actionType = actionObj["type"]?.jsonPrimitive?.contentOrNull ?: return null
        val intent = runCatching { json.decodeFromString<VoiceIntentType>("\"$intentStr\"") }
            .getOrDefault(VoiceIntentType.UNKNOWN)
        val gesture = obj["gesture"]?.jsonObject?.let { g ->
            val name = g["gesture"]?.jsonPrimitive?.contentOrNull ?: return@let null
            GestureState(gesture = name, duration = g["duration"]?.jsonPrimitive?.intOrNull)
        }
        return VoiceWSIncoming.IntentAction(
            intent = intent, text = text,
            action = VoiceAction(type = actionType, payload = actionObj["payload"]?.jsonObject),
            confidence = obj["confidence"]?.jsonPrimitive?.doubleOrNull, gesture = gesture,
        )
    }

    private fun strVal(obj: JsonObject, key: String): String? =
        obj[key]?.jsonPrimitive?.contentOrNull

    private fun decodeBase64Val(obj: JsonObject, key: String): ByteArray? {
        val b64 = strVal(obj, key) ?: return null
        return runCatching { android.util.Base64.decode(b64, android.util.Base64.DEFAULT) }.getOrNull()
    }

    companion object {
        private const val NORMAL_CLOSURE_CODE = 1000
        private const val NORMAL_CLOSURE_REASON = "Client disconnect"
    }
}
